compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         (expected, Some(err))
     }
 
+    fn point_inference_types(&self, err: &mut Diagnostic, expr: &hir::Expr<'_>) -> bool {
+        let tcx = self.tcx;
+        let map = self.tcx.hir();
+
+        // Hack to make equality checks on types with inference variables and regions useful.
+        struct TypeEraser<'tcx> {
+            tcx: TyCtxt<'tcx>,
+        }
```

Comment:
This can use a [`BottomUpFolder`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/fold/struct.BottomUpFolder.html) probably

Reply:
Nice! I hadn't seen that when it was added, I'm guessing to support TAIT?